rc/tools/lint.kak

```diff
@@ -1,113 +1,339 @@
+                        kakquote(line "|" flag)
+                }
+
+                for (line in messages_by_line) {
+                    msg = messages_by_line[line]
+
+                    print "set-option -add " \
+                        kakquote("buffer=" file) " " \
+                        "lint_messages " \
+                        kakquote(line "|" msg)
+                }
+
+                print "set-option " \
+                    kakquote("buffer=" file) " " \
+                    "lint_error_count " \
+                    error_count
+                print "set-option " \
+                    kakquote("buffer=" file) " " \
+                    "lint_warning_count " \
+                    warning_count
             }
-        ' "$dir"/stderr | kak -p "$kak_session"
+        ' "$dir"/result | kak -p "$kak_session"
+
+        # Send any linting errors to the debug buffer,
+        # for visibility.
+        if [ -s "$dir"/stderr ]; then
+            # Errors were detected!"
+            printf "echo -debug Linter errors: <<<\n"
+            while read -r LINE; do
+                printf "echo -debug %s\n" "$(kakquote "  $LINE")"
+            done < "$dir"/stderr
+            printf "echo -debug >>>\n"
+            # FIXME: When #3254 is fixed, this can become a "fail"
+            printf "eval -client %s echo -markup {Error}%s\n" \
+                "$kak_client" \
+                "lint failed, see *debug* for details"
+        else
+            # No errors detected, show the results.
+            printf "eval -client %s lint-show-counters" \
+                "$kak_client"
+        fi | kak -p "$kak_session"
+
+        # We are done here. Send the results to Kakoune,
+        # and clean up.
+        cat "$dir"/result > "$dir"/fifo
+        rm -rf "$dir"
+
+        } & ) >"$dir"/stderr 2>&1 </dev/null
+    }
+}
+
+define-command \
+    -params 0..2 \
+    -docstring %{
+        lint-selections [<switches>]: Check each selection with a linter.
+
+        Switches:
+            -command <cmd>      Use the given linter.
+                                If not given, the lintcmd option is used.
+    } \
+    lint-selections \
+%{
+    evaluate-commands -draft %{
+        # Make sure all the selections are "forward" (anchor before cursor)
+        execute-keys <a-:>
 
-        cut -d: -f2- "$dir"/stderr | awk -v bufname="${kak_bufname}" '
-            /^[1-9][0-9]*:[1-9][0-9]*:/ {
-                print bufname ":" $0
+        # Make sure the selections are in document order.
+        evaluate-commands %sh{
+            printf "select "
+            printf "%s\n" "$kak_selections_desc" |
+                tr ' ' '\n' |
+                sort -n -t. |
+                tr '\n' ' '
+        }
+
+        evaluate-commands %sh{
+            # This is going to come in handy later.
+            kakquote() {
+                printf "'"
+                printf "%s" "$*" | sed "s/'/''/g"
+                printf "'"
```

Comment:
It could, but it's actually a bit slower than the current implementation. I've (recently) filed #3340 about making a fast implementation automatically available.

Reply:
We can shelve that for the time being, to avoid discussing the same topic in two different threads.

But I think making that helper that is duplicated in a couple shell scopes as small as possible is more important that making it the fastest possible, as long as it's not taking *too long* (term to be defined) to run.

Reply:
I found a different way to shorten the function into one line. I agree, it does look tidier.